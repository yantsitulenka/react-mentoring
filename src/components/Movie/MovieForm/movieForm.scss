.add-movie-form {
  background-color: #232323;
  padding: 2rem 3rem;
  color: #ffffff;
  display: flex;
  flex-direction: column;
  font-family: 'Montserrat', sans-serif;
  position: relative;

  &__title {
    font-size: 30px;
    font-weight: 200;
    letter-spacing: 2px;
    text-transform: uppercase;
  }

  &__close {
    position: absolute;
    right: 0;
    margin: 15px;
    top: 0;
    font-size: 24px;
    cursor: pointer;
  }

  &__fields {
    display: flex;
    flex-direction: column;
    position: relative;

    &-validation {
      position: absolute;
      font-size: 12px;
      top: 80px;
      color: #f65261;
      font-weight: 200;
    }

    & > label {
      color: #f65261;
      text-transform: uppercase;
      padding: 5px 0;
      font-size: 17px;
      letter-spacing: 1px;
    }

    & > input {
      width: 400px;
      color: #ffffff;
      background-color: #555555;
      border: none;
      padding: 10px;
      font-size: 17px;
      border-radius: 3px;
      margin: 5px 0 30px;

      &:focus {
        border: 2px solid #f65261;
        outline: none;
        padding: 8px;
      }
    }
  }

  &__selector {
    width: 400px;
    color: #ababab;
    background-color: #555555;
    position: relative;
    border: none;
    padding: 10px;
    font-size: 17px;
    border-radius: 3px;
    margin: 5px 0 30px;
    cursor: pointer;

    &-list {
      display: flex;
      flex-direction: column;
      background-color: #232323;
      position: absolute;
      width: calc(100% - 20px);
      padding: 10px;
      box-shadow: 0 0 10px rgba(0,0,0,0.5);
      top: 80px;
      z-index: 2;

      label {
        cursor: pointer;
      }
    }
  }

  &__id {
    margin-bottom: 20px;
    color: #555555;
  }

  &__buttons {
    display: flex;
    justify-content: space-between;

    .button {
      padding: 17px 72px;
      border-radius: 8px;
      border: none;
      color: #fff;
      background-color: #f65261;
      cursor: pointer;
      text-transform: uppercase;
      font-weight: 500;
      letter-spacing: 1.5px;
      font-family: 'Montserrat', sans-serif;

      &.secondary {
        border: 1px solid #f65261;
        background: none;
        padding: 15px 70px;;
      }
    }
  }

  .hidden {
    display: none;
  }
}
