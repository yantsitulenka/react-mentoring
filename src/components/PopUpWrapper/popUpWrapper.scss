.pop-up-wrapper {
  position: fixed;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1;
  background-color: rgba(0,0,0,.8);
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  &__error {
    color: #f65261;
    margin-bottom: 1rem;
  }
}
