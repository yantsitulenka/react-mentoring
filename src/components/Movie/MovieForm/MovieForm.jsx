import React, { useState } from 'react';
import { connect } from 'react-redux';
import {
  Formik, Form, Field, ErrorMessage,
} from 'formik';
import * as Yup from 'yup';
import './movieForm.scss';
import {
  closeEditMovieForm, editMovieAsync, closeAddMovieForm, addMovieAsync,
} from '../../../actions';

const SignUpSchema = Yup.object().shape({
  title: Yup.string()
    .min(2, 'Must be longer than 2 characters')
    .required('This field is required'),
  release_date: Yup.string()
    .min(2, 'Must be longer than 2 characters')
    .required('This field is required'),
  poster_path: Yup.string()
    .min(2, 'Must be longer than 2 characters')
    .required('This field is required'),
  overview: Yup.string()
    .min(2, 'Must be longer than 2 characters')
    .required('This field is required'),
  runtime: Yup.string()
    .min(2, 'Must be longer than 2 characters')
    .required('This field is required'),
});

const MovieForm = ({
  tempMovie, closeEditMovieForm, closeAddMovieForm, submitAddMovie, submitEditMovie,
}) => {
  const isEditForm = !!tempMovie.title;
  const [showGenreList, setShowGenreList] = useState(false);

  return (
    <div className="add-movie-form">
      <div className="add-movie-form__close" />
      <span
        className="add-movie-form__close"
        onClick={() => (isEditForm ? closeEditMovieForm() : closeAddMovieForm())}
        role="button"
      >
        &#10006;
      </span>
      <h2 className="add-movie-form__title">Add Movie</h2>

      <Formik
        initialValues={{
          title: tempMovie?.title || '',
          release_date: tempMovie?.release_date || '',
          poster_path: tempMovie?.poster_path || '',
          overview: tempMovie?.overview || '',
          runtime: tempMovie?.runtime || '',
          genres: tempMovie?.genres || '',
        }}
        validationSchema={SignUpSchema}
        onSubmit={
          (val) => (isEditForm
            ? submitEditMovie({ ...tempMovie, ...val, runtime: parseInt(val.runtime, 10) })
            : submitAddMovie({ ...val, runtime: parseInt(val.runtime, 10) }))
        }
        render={({ resetForm }) => (
          <Form>
            <div className="add-movie-form__fields">
              <label htmlFor="title">Title</label>
              <Field name="title" placeholder="Title" type="text" />
              <ErrorMessage
                name="title"
                component="div"
                className="add-movie-form__fields-validation"
              />
            </div>

            <div className="add-movie-form__fields">
              <label htmlFor="release_date">Release Date</label>
              <Field name="release_date" placeholder="2000-01-01" type="text" />
              <ErrorMessage
                name="release_date"
                component="div"
                className="add-movie-form__fields-validation"
              />
            </div>

            <div className="add-movie-form__fields">
              <label htmlFor="poster_path">POster path</label>
              <Field name="poster_path" type="text" />
              <ErrorMessage
                name="poster_path"
                component="div"
                className="add-movie-form__fields-validation"
              />
            </div>

            <div className="add-movie-form__fields">
              <label htmlFor="overview">Overview</label>
              <Field name="overview" type="text" />
              <ErrorMessage
                name="overview"
                component="div"
                className="add-movie-form__fields-validation"
              />
            </div>

            <div className="add-movie-form__fields">
              <label htmlFor="genre">GenreL</label>
              <div id="genre" className="add-movie-form__selector" onClick={() => setShowGenreList(!showGenreList)}>Select Genre</div>
              <div className={`add-movie-form__selector-list ${showGenreList ? '' : 'hidden'}`} role="group" aria-labelledby="checkbox-group">
                <label>
                  <Field type="checkbox" name="genres" value="crime" />
                  Crime
                </label>
                <label>
                  <Field type="checkbox" name="genres" value="documentary" />
                  Documentary
                </label>
                <label>
                  <Field type="checkbox" name="genres" value="horror" />
                  Horror
                </label>
                <label>
                  <Field type="checkbox" name="genres" value="comedy" />
                  Comedy
                </label>
              </div>
            </div>

            <div className="add-movie-form__fields">
              <label htmlFor="runtime">Runtime</label>
              <Field name="runtime" type="text" />
              <ErrorMessage
                name="runtime"
                component="div"
                className="add-movie-form__fields-validation"
              />
            </div>

            <div className="add-movie-form__buttons">
              <button type="button" className="button secondary" onClick={resetForm}>Reset</button>
              <button type="submit" className="button asd">Submit</button>
            </div>
          </Form>
        )}
      />
    </div>
  );
};

const mapStateToProps = (state) => ({
  tempMovie: state.movieToEdit,
});

const mapDispatchToProps = (dispatch) => ({
  closeEditMovieForm: () => dispatch(closeEditMovieForm()),
  closeAddMovieForm: () => dispatch(closeAddMovieForm()),
  submitEditMovie: (movie) => dispatch(editMovieAsync(movie)),
  submitAddMovie: (movie) => dispatch(addMovieAsync(movie)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MovieForm);
