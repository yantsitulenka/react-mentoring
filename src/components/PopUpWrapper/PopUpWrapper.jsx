import React from 'react';
import { connect } from 'react-redux';
import './popUpWrapper.scss';

const PopUpWrapper = ({ children, apiError }) => (
  <div className="pop-up-wrapper">
    {
      apiError ? (
        <div className="pop-up-wrapper__error"> System error, please try again latter</div>
      ) : ''
    }
    {children}
  </div>
);

const mapStateToProps = (state) => ({
  apiError: state.apiError,
});

export default connect(
  mapStateToProps,
)(PopUpWrapper);
