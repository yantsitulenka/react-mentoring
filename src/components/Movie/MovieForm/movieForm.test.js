import React from 'react';
import rerender from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import {
  render, fireEvent, getByText,
} from '@testing-library/react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MovieForm from './MovieForm';
import PopUpWrapper from '../../PopUpWrapper/PopUpWrapper';

configure({ adapter: new Adapter() });

describe('when MovieForm', () => {
  test('then snapshot created for add movie', () => {
    const initialStore = {
      movieToEdit: {},
    };
    const store = configureStore([])(initialStore);
    store.dispatch = jest.fn();
    const movieFilterComponent = rerender.create(
      <Provider store={store}>
        <PopUpWrapper>
          <MovieForm />
        </PopUpWrapper>
      </Provider>,
    );

    const tree = movieFilterComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('then snapshot created for edit movie', () => {
    const initialStore = {
      movieToEdit: {
        id: 337167,
        title: 'Fifty Shades Freed',
        tagline: "Don't miss the climax",
        vote_average: 6.1,
        vote_count: 1195,
        release_date: '2018-02-07',
        poster_path: 'https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg',
        overview: 'Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.',
        budget: 55000000,
        revenue: 136906000,
        genres: [
          'Drama',
          'Romance',
        ],
        runtime: 106,
      },
    };
    const store = configureStore([])(initialStore);
    store.dispatch = jest.fn();
    const movieFilterComponent = rerender.create(
      <Provider store={store}>
        <PopUpWrapper>
          <MovieForm />
        </PopUpWrapper>
      </Provider>,
    );

    const tree = movieFilterComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test click on genre', () => {
    const initialStore = {
      movieToEdit: {},
    };
    const store = configureStore([])(initialStore);
    const { container } = render(
      <Provider store={store}>
        <PopUpWrapper>
          <MovieForm />
        </PopUpWrapper>
      </Provider>,
    );
    const countValue = getByText(container, 'Select Genre');
    const list = container.querySelector('.add-movie-form__selector-list');
    expect(list.classList.contains('hidden')).toBe(true);
    fireEvent.click(countValue);
    expect(list.classList.contains('hidden')).toBe(false);
  });

});
