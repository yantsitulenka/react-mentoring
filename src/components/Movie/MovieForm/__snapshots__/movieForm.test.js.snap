// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="pop-up-wrapper"
>
  
  <div
    className="add-movie-form"
  >
    <div
      className="add-movie-form__close"
    />
    <span
      className="add-movie-form__close"
      onClick={[Function]}
      role="button"
    >
      ✖
    </span>
    <h2
      className="add-movie-form__title"
    >
      Add Movie
    </h2>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Title"
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="release_date"
        >
          Release Date
        </label>
        <input
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2000-01-01"
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="poster_path"
        >
          POster path
        </label>
        <input
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="genre"
        >
          GenreL
        </label>
        <div
          className="add-movie-form__selector"
          id="genre"
          onClick={[Function]}
        >
          Select Genre
        </div>
        <div
          aria-labelledby="checkbox-group"
          className="add-movie-form__selector-list hidden"
          role="group"
        >
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="crime"
            />
            Crime
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="documentary"
            />
            Documentary
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="horror"
            />
            Horror
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="comedy"
            />
            Comedy
          </label>
        </div>
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__buttons"
      >
        <button
          className="button secondary"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
        <button
          className="button asd"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`when MovieForm then snapshot created for add movie 1`] = `
<div
  className="pop-up-wrapper"
>
  
  <div
    className="add-movie-form"
  >
    <div
      className="add-movie-form__close"
    />
    <span
      className="add-movie-form__close"
      onClick={[Function]}
      role="button"
    >
      ✖
    </span>
    <h2
      className="add-movie-form__title"
    >
      Add Movie
    </h2>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Title"
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="release_date"
        >
          Release Date
        </label>
        <input
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2000-01-01"
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="poster_path"
        >
          POster path
        </label>
        <input
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="genre"
        >
          GenreL
        </label>
        <div
          className="add-movie-form__selector"
          id="genre"
          onClick={[Function]}
        >
          Select Genre
        </div>
        <div
          aria-labelledby="checkbox-group"
          className="add-movie-form__selector-list hidden"
          role="group"
        >
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="crime"
            />
            Crime
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="documentary"
            />
            Documentary
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="horror"
            />
            Horror
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="comedy"
            />
            Comedy
          </label>
        </div>
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="add-movie-form__buttons"
      >
        <button
          className="button secondary"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
        <button
          className="button asd"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`when MovieForm then snapshot created for edit movie 1`] = `
<div
  className="pop-up-wrapper"
>
  
  <div
    className="add-movie-form"
  >
    <div
      className="add-movie-form__close"
    />
    <span
      className="add-movie-form__close"
      onClick={[Function]}
      role="button"
    >
      ✖
    </span>
    <h2
      className="add-movie-form__title"
    >
      Add Movie
    </h2>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Title"
          type="text"
          value="Fifty Shades Freed"
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="release_date"
        >
          Release Date
        </label>
        <input
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2000-01-01"
          type="text"
          value="2018-02-07"
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="poster_path"
        >
          POster path
        </label>
        <input
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins."
        />
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="genre"
        >
          GenreL
        </label>
        <div
          className="add-movie-form__selector"
          id="genre"
          onClick={[Function]}
        >
          Select Genre
        </div>
        <div
          aria-labelledby="checkbox-group"
          className="add-movie-form__selector-list hidden"
          role="group"
        >
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="crime"
            />
            Crime
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="documentary"
            />
            Documentary
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="horror"
            />
            Horror
          </label>
          <label>
            <input
              checked={false}
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="comedy"
            />
            Comedy
          </label>
        </div>
      </div>
      <div
        className="add-movie-form__fields"
      >
        <label
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value={106}
        />
      </div>
      <div
        className="add-movie-form__buttons"
      >
        <button
          className="button secondary"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
        <button
          className="button asd"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
